### ISSUE TITLE
Can't change autoloaded arrays that have a defined type

### ISSUE THREAD

**Issue by morfyy** [godotengine/godot#56473]
### Godot version

3.4

### System information

Windows 10

### Issue description

It seems like that when you declare an array in an autoloaded script and give it a specific type like PoolIntArray, you can't use methods like remove() or append() on it. 

### Steps to reproduce

Autoload script:
```
extends Node
var array:PoolIntArray = [1, 2, 3]
```

Local script:
```
g.array.remove(0)
```

Result is that nothing is removed from the array.

This however still works:
```
g.array = [2, 3]
```

If the array had been defined with no type or the type `Array` or if it was declared inside the local script, then this problem wouldnt occur.

**Comment by cdemirer** [comment]
Hmm, it seems this is related to #15265.

It seems Pool Array has design issues (as a mutable "immutable"), which are remedied in 4.0 with Packed Array. I guess it's not a good idea to use Pool Array in the general case, and only with great care when needed.

**Comment by fire540** [comment]
I don't think this is technically a bug because pool arrays are passed by value (like int, string, float, etc) and not reference (like Array and Dictionary). I generally don't use pool arrays for this reason, as this has caused many bugs in my code before. In your case, I would just use a regular Array.

All pool array types have this note in them:
![image](https://user-images.githubusercontent.com/67974470/148159123-1dc1a11b-7310-444b-b32d-6b28e24aba70.png)

**Comment by cdemirer** [comment]
I agree that this is technically not a bug. However, I don't think that it's sufficiently documented, since "passed by value" doesn't clarify this behavior well. It's very natural to expect the other behavior even with the knowledge that it passes by value.

In C++, std::vector (or even Godot's Vector) can be considered to "pass by value". But when you call a member function of a Vector through a chain of `.`s, you work on the existing object, without creating a temporary copy. This is different in GDScript, and mentioning this difference would make the documentation more clear in my opinion.

**Comment by Calinou** [comment]
Pool*Array is passed by value in `3.x`, but Packed*Array is passed by reference in `3.x` unless there is a need to pass it by value (in which case it's passed by value).

**Comment by cdemirer** [comment]
> As I understand it, Pool*Array is passed by value in `3.x`, but Packed*Array is passed by reference in `master` unless there is a need to pass it by value. In this case, Packed*Array is automatically copied (and is therefore passed by value from the user's perspective).

I believe you're correct (assuming you meant Pool*Array in the last sentence).

---

The main problem here is that the subtle fact that the passing behavior is relevant in the first place (which is the thing that would benefit from documentation I think).

`object.member.member_function()` in C++ doesn't lead to a "passing" at all. It acts on the existing object identified by `object.member`.
However, in GDScript, this leads to the creation of a temporary object, which makes the passing behavior relevant.
When evaluating `object.member.member_function()`, it is parsed as a function call:
Base: `object.member`
Function: `member_function`
Before the call is evaluated, first the base is evaluated, which creates a temporary, and assigns it the result of something like `get_member(object, member)`. This is the reason it has to "pass" (by anything) in the first place, and is very subtle.

While this is true for member function calls on members of all types, I think Pool*Array would benefit the most from this documentation, as it's the only one that is affected by this. All the other types are either "really" immutable or passed by reference, which lets the user ignore all of this.

**Comment by akien-mga** [comment]
Duplicate of #15265. That's a design limitation in 3.x, fixed in 4.0.
